### PR TITLE
Composition verification for external sources

### PR DESCRIPTION
**Pull request by d-montgomery**
This PR adds a test to the `EB_ODEQty` regression test, namely `composition-test-2d.inp`, which verifies the implementation of user-defined external sources for density $\rho$ and species mass fractions $Y_m$.  The initial mass fractions are set to $Y_{CO2} = 0.15$, $Y_{N2} = 0.25$, and $Y_{AR} = 0.6$.

After $t_s$ seconds, CO2 is added at a rate $S_{ext,\rho Y_{CO2}}$, which also causes the density to increase at a rate of $S_{ext,\rho} = S_{ext,\rho Y_{CO2}}$.

The results can be plotted against the analytical solutions in `plot-comps.py` (see figure below) and tested in the CI via `test.py`, where errors must be less than a specified maximum error threshold.  This was done by adding `add_test_rt` to `CMakeLists`, which runs any python script called `test.py` in the specified case directory.

Additionally, this PR fixes a minor bug in `problem_modify_ext_sources`, where time is passed in via `a_timestamp_old`  instead of `a_timestamp_new` due to the explicit treatment of the external source terms.

composition-test

### CODE REVIEW

**Comment by d-montgomery** [comment]
Documenting the problem with ubuntu-24.04 EB-ON CI test:
The new `test.py` test for `EB_ODEQty/composition-test-2d.inp` passes during the macOS-latest EB-ON test, but fails during the ubuntu-24.04 EB-ON test.  As shown in commit [#c7a863a](https://github.com/AMReX-Combustion/PeleLMeX/pull/443/commits/c7a863abde753766d7ea01133d9882e864a54d26), the [macOS](https://github.com/AMReX-Combustion/PeleLMeX/actions/runs/12131256303/job/33823088561#step:12:38) and [ubuntu](https://github.com/AMReX-Combustion/PeleLMeX/actions/runs/12131256303/job/33823087672#step:12:39) tests run the associated PeleLMeX executable with four processors.  The ubuntu test appears to be running the PeleLMeX executable four times while the macOS test runs the test in parallel.  The `temporals/tempExtremas` data for each test are included in the images  below.

Data read in during macOS test:
macos-test

Data read in during the ubuntu test (4 processors each adding to tempExtremas):
ubuntu-test

File: Tests/CMakeLists.txt [hunk 1]
```diff
@@ -110,6 +110,14 @@ function(add_test_rev TEST_NAME TEST_EXE_DIR)
     set_tests_properties(${TEST_NAME} PROPERTIES LABELS "regression;verification;no-ci")
 endfunction(add_test_rev)
 
+# Regression test with case-specific verification test.py
+function(add_test_rt TEST_NAME TEST_EXE_DIR)
+    setup_test()
+    set(RUNTIME_OPTIONS "amr.max_step=10 peleLM.do_temporals=1 peleLM.do_extremas=1 peleLM.temporal_int=1 ${RUNTIME_OPTIONS}")
+    add_test(${TEST_NAME} sh -c "${CURRENT_TEST_EXE} ${MPIEXEC_POSTFLAGS} ${CURRENT_TEST_BINARY_DIR}/${TEST_NAME}.inp ${RUNTIME_OPTIONS} > ${TEST_NAME}.log ${SAVE_GOLDS_COMMAND} ${FCOMPARE_COMMAND} && nosetests ${CURRENT_TEST_BINARY_DIR}/test.py")
```
Comment (d-montgomery) [inline Tests/CMakeLists.txt]: In order to pass the ubuntu CI test, I had to remove `${MPI_COMMANDS}` from `add_test()`.  This is a work around for now as I don't think the problem is related to this PR, but more so related to the MPI commands in the ubuntu build.

**Comment by jrood-nrel** [comment]
Let's see if MPI works better if we step ubuntu back https://github.com/AMReX-Combustion/PeleLMeX/pull/446 until I figure what happened to MPI in the latest image.

**Review by jrood-nrel — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -84,7 +84,7 @@ jobs:
     runs-on: ${{matrix.os}}
     strategy:
       matrix:
-        os: [ubuntu-24.04, macos-latest]
+        os: [ubuntu-latest, macos-latest]
```
Comment (jrood-nrel) [inline .github/workflows/ci.yml]: Let's change to `ubuntu-22.04` here.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -93,7 +93,7 @@ jobs:
             comp: llvm
             procs: $(sysctl -n hw.ncpu)
             ccache_cache: /Users/runner/Library/Caches/ccache
-          - os: ubuntu-24.04
+          - os: ubuntu-latest
```
Comment (jrood-nrel) [inline .github/workflows/ci.yml]: Let's change to `ubuntu-22.04` here.

**Comment by jrood-nrel** [comment]
So we can use `ubuntu-24.04` which is also now `ubuntu-latest` on everything except for when we run with MPI. Looks like we need `ubuntu-22.04` to run with MPI. I'm trying to figure out what happened to MPI on the new image.

**Review by jrood-nrel — APPROVED** [review]

**Comment by d-montgomery** [comment]
@jrood-nrel, any idea why there are expected tests for CPU-CMake ubuntu-24.04 in the testing queue?

**Comment by jrood-nrel** [comment]
Those have to be updated after this is merged. It is because the required tests were renamed.

File: Exec/RegTests/EB_ODEQty/eb-odeqty-3d.inp
Comment (baperry2) [inline Exec/RegTests/EB_ODEQty/eb-odeqty-3d.inp:1]: Can we just delete this file? It looks like we don't use the 3D version in the CI and this file is also ~identical to the 2D one anyway

File: Exec/RegTests/EB_ODEQty/PeleLMeX_ProblemSpecificFunctions.cpp [hunk 4]
```diff
@@ -45,10 +45,32 @@ problem_modify_ext_sources(
...
+      if (prob_parm.ode_qty_test) {
+        for (int n = 0; n < NUM_ODE; n++) {
+          // Source terms for ODE qty test
+
+          Real B_n = state_old_arr[box_no](i, j, k, FIRSTODE + n);
+          Real src = prob_parm.ode_srcstrength * pow(10.0, n + 1) * B_n;
+          ext_src_arr[box_no](i, j, k, FIRSTODE + n) += src;
+        }
+        // Ignore time as it is only used if composition_test = 1
+        amrex::ignore_unused(time);
```
Comment (baperry2) [inline Exec/RegTests/EB_ODEQty/PeleLMeX_ProblemSpecificFunctions.cpp]: This ignore unused and the one below shouldn't be needed. ignore-Unused should only be needed when certain variables are excluded by compile time macros, not runtime conditionals

File: Exec/RegTests/EB_ODEQty/eb-odeqty-3d.inp
Comment (d-montgomery) [reply Exec/RegTests/EB_ODEQty/eb-odeqty-3d.inp:1]: `eb-odeqty-3d.inp` is used in the [EB-ON tests](https://github.com/AMReX-Combustion/PeleLMeX/actions/runs/12357654428/job/34486432543#step:12:45) and was included in [PR #427](https://github.com/AMReX-Combustion/PeleLMeX/pull/427) for consistency with other CI tests that run in 2D and 3D.  We can remove it and have a single `eb-odeqty.inp` run in 2D and 3D if you'd like.

Comment (baperry2) [reply Exec/RegTests/EB_ODEQty/eb-odeqty-3d.inp:1]: Oh, I see now, you can ignore this comment.

**Review by baperry2 — APPROVED** [review]
Looks good, just the minor change for ignore-unused.

File: Exec/RegTests/EB_ODEQty/PeleLMeX_ProblemSpecificFunctions.cpp (hunk 4 above)
Comment (d-montgomery) [reply Exec/RegTests/EB_ODEQty/PeleLMeX_ProblemSpecificFunctions.cpp]: Thanks for clarifying.  I made the update and merged the most recent branch.

**Pull request merged**